Fix and Test Conditional Validator Behavior in FormField

EdgarJan opened the pull request:
In the FormField widget, if a validator is initially set (and validation fails), then subsequently the validator is set to null, the form incorrectly retains its error state. This is not expected behavior as removing the validator should clear any validation errors.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

EdgarJan committed:
Clear FormField errorText when validator is set to null

EdgarJan committed:
Add test for conditional validator behavior in FormField

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -940,4 +940,61 @@ void main() {
+  final GlobalKey<FormState> formKey = GlobalKey<FormState>();
+  bool useValidator = false;
+
+  String? errorText(String? value) {
```
Nit: I feel like this would be better named `validator`.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -940,4 +940,61 @@ void main() {
+    );
+  }
+
+  // Start with no validator
```
Periods at the end of these comments.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -940,4 +940,61 @@ void main() {
     expect(fieldKey.currentState!.hasInteractedByUser, isFalse);
   });
+
+testWidgets('Validator is nullified and error text behaves accordingly',
```
Indent this whole test by 2 spaces.

justinmc commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -940,4 +940,61 @@ void main() {
+
+  // Now use the validator
+  useValidator = true;
+  await tester.pumpWidget(buildTestApp());
```
Nit: Would it be any better to do this test with setState instead of re-pumping an entire widget tree?  I'm not sure if it would really matter here or not, but it might be more reassuring.

You could use StatefulBuilder [like this](https://github.com/flutter/flutter/blob/00a99d3bc09bed0433319f05f5161557d11039f8/packages/flutter/test/widgets/form_test.dart#L471).

justinmc reviewed (commented):
I think you're right about this, thanks for fixing it!  Just some minor comments.

EdgarJan committed:
Address review feedback: Refactor tests

Address review comments:
- Rename 'errorText' function to 'validator'
- Use StatefulBuilder for better state management in tests
- Add periods to test comments for clarity

Piinks commented:
Welcome @EdgarJan! thanks for contributing. Does this PR close an open issue?

EdgarJan commented:
@Piinks No, I ran into an issue myself and fixed it without opening anything. If it's mandatory I can open issue and explain the problem.

Piinks commented:
> If it's mandatory I can open issue and explain the problem.

For this case, I think we can go without. Just making sure we closed/referenced it if it already existed.

Piinks reviewed (approved):
This LGTM, justinmc is OOO for now. I'll find another reviewer to give this a second stamp.

LongCatIsLooong commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -524,6 +524,8 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
     if (widget.validator != null) {
       _errorText.value = widget.validator!(_value);
+    } else {
+      _errorText.value = null;
```
nit: this could just be `_errorText.value = widget.validator?.call(_value);`
